=== modified file 'src/auth/digest/Config.cc'
--- src/auth/digest/Config.cc	2015-09-27 17:36:15 +0000
+++ src/auth/digest/Config.cc	2015-09-28 05:50:07 +0000
@@ -646,7 +646,7 @@
         if (tmplink->data == nonce) {
             dlinkDelete(tmplink, &digest_user->nonces);
             authDigestNonceUnlink(static_cast < digest_nonce_h * >(tmplink->data));
-            dlinkNodeDelete(tmplink);
+            delete tmplink;
             link = NULL;
         }
     }
@@ -676,7 +676,7 @@
     if (node)
         return;
 
-    node = dlinkNodeNew();
+    node = new dlink_node;
 
     dlinkAddTail(nonce, node, &digest_user->nonces);
 

=== modified file 'src/auth/digest/User.cc'
--- src/auth/digest/User.cc	2015-01-13 07:25:36 +0000
+++ src/auth/digest/User.cc	2015-09-28 05:50:07 +0000
@@ -32,7 +32,7 @@
         dlinkDelete(tmplink, &nonces);
         authDigestNoncePurge(static_cast < digest_nonce_h * >(tmplink->data));
         authDigestNonceUnlink(static_cast < digest_nonce_h * >(tmplink->data));
-        dlinkNodeDelete(tmplink);
+        delete tmplink;
     }
 }
 

=== modified file 'src/dlink.cc'
--- src/dlink.cc	2015-01-13 07:25:36 +0000
+++ src/dlink.cc	2015-09-28 05:53:46 +0000
@@ -9,23 +9,12 @@
 #include "squid.h"
 #include "dlink.h"
 
-/* dlink_node use explicit alloc()/freeOne()
- * XXX: convert to MEMPROXY_CLASS() API
- */
-#include "mem/Pool.h"
-
 dlink_list ClientActiveRequests;
 
-MemAllocator *dlink_node_pool = NULL;
-
 dlink_node *
 dlinkNodeNew()
 {
-    if (dlink_node_pool == NULL)
-        dlink_node_pool = memPoolCreate("Dlink list nodes", sizeof(dlink_node));
-
-    /* where should we call delete dlink_node_pool;dlink_node_pool = NULL; */
-    return (dlink_node *)dlink_node_pool->alloc();
+	return new dlink_node;
 }
 
 /** The node needs to be unlinked FIRST */
@@ -35,7 +24,7 @@
     if (m == NULL)
         return;
 
-    dlink_node_pool->freeOne(m);
+    delete m;
 }
 
 void

=== modified file 'src/dlink.h'
--- src/dlink.h	2015-01-13 07:25:36 +0000
+++ src/dlink.h	2015-09-28 05:53:46 +0000
@@ -9,11 +9,13 @@
 #ifndef SQUID_DLINK_H
 #define SQUID_DLINK_H
 
+#include "mem/forward.h"
+
 class dlink_node
 {
-
+	MEMPROXY_CLASS(dlink_node);
 public:
-    dlink_node() : data(NULL), prev(NULL), next(NULL) {}
+    dlink_node() : data(nullptr), prev(nullptr), next(nullptr) {}
 
     void *data;
     dlink_node *prev;
@@ -35,8 +37,6 @@
 void dlinkAddAfter(void *, dlink_node *, dlink_node *, dlink_list *);
 void dlinkAddTail(void *data, dlink_node *, dlink_list *);
 void dlinkDelete(dlink_node * m, dlink_list * list);
-void dlinkNodeDelete(dlink_node * m);
-dlink_node *dlinkNodeNew(void);
 
 #endif /* SQUID_DLINK_H */
 

