diff --git a/src/ssl/PeekingPeerConnector.cc b/src/ssl/PeekingPeerConnector.cc
index 0d680d88e..d6fc13063 100644
--- a/src/ssl/PeekingPeerConnector.cc
+++ b/src/ssl/PeekingPeerConnector.cc
@@ -295,6 +295,9 @@ Ssl::PeekingPeerConnector::noteNegotiationError(const int result, const int ssl_
             // if (const auto spliceResumed = true) {
             bypassValidator = true;
             // } // else fall through to find a matching ssl_bump action (with limited info)
+        } else if (ssl_lib_error == 0x14094410) {
+            bypassValidator = true;
+            debugs(83, 3, "Error " << ssl_lib_error << " detected; bypassing");
         }
 
         if (bypassValidator) {

