=== modified file 'configure.ac'
--- configure.ac	2016-01-04 14:39:06 +0000
+++ configure.ac	2016-01-05 08:14:39 +0000
@@ -458,10 +458,13 @@
 fi
 
 ## check for atomics library before anything that might need it
+# AC_SEARCH_LIBS pollutes LIBS
+SQUID_STATE_SAVE(LIBATOMIC)
 AC_SEARCH_LIBS([__atomic_load_8],[atomic])
-if test "x$ac_cv_search___atomic_load_8" = "-latomic"; then
+if test "x$ac_cv_search___atomic_load_8" = "x-latomic"; then
   ATOMICLIB="-latomic"
 fi
+SQUID_STATE_ROLLBACK(LIBATOMIC)
 AC_SUBST(ATOMICLIB)
 
 AC_SEARCH_LIBS([shm_open], [rt])

