=== modified file 'src/ip/Intercept.cc'
--- src/ip/Intercept.cc 2015-08-16 07:56:20 +0000
+++ src/ip/Intercept.cc 2015-09-22 09:30:04 +0000
@@ -87,6 +87,10 @@
 #if LINUX_NETFILTER
 /* <climits> must be before including netfilter_ipv4.h */
 #include <climits>
+#ifdef _LIBC_COMPAT_H
+/* workaround for broken cross-include detection in linux/libc-compat.h */
+#undef _LIBC_COMPAT_H
+#endif
 #include <linux/if.h>
 #include <linux/netfilter_ipv4.h>
 #if HAVE_LINUX_NETFILTER_IPV6_IP6_TABLES_H


