=== modified file 'src/http/RegisteredHeadersHash.cci'
--- src/http/RegisteredHeadersHash.cci	2015-09-06 00:12:11 +0000
+++ src/http/RegisteredHeadersHash.cci	2015-09-22 15:05:16 +0000
@@ -40,6 +40,10 @@
  * contributions from numerous individuals and organizations.
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
+
+/* the register keyword is deprecated in c++11, and some compilers barf on it. */
+#define register
+
 #line 24 "RegisteredHeadersHash.gperf"
 struct HeaderTableRecord;
         enum

=== modified file 'src/http/RegisteredHeadersHash.gperf'
--- src/http/RegisteredHeadersHash.gperf	2015-09-05 18:52:17 +0000
+++ src/http/RegisteredHeadersHash.gperf	2015-09-22 15:05:02 +0000
@@ -8,6 +8,9 @@
  * contributions from numerous individuals and organizations.
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
+
+/* the register keyword is deprecated in c++11, and some compilers barf on it. */
+#define register
 %}
 %language=C++
 %define hash-function-name HttpHeaderHash

