=== modified file 'src/client_side.cc'
--- src/client_side.cc	2015-03-21 07:53:13 +0000
+++ src/client_side.cc	2015-04-30 04:03:52 +0000
@@ -3968,7 +3968,7 @@
                 if (sslServerBump && (sslServerBump->act.step1 == Ssl::bumpPeek || sslServerBump->act.step1 == Ssl::bumpStare)) {
                     doPeekAndSpliceStep();
                     SSL *ssl = fd_table[clientConnection->fd].ssl;
-                    bool ret = Ssl::configureSSLUsingPkeyAndCertFromMemory(ssl, reply_message.getBody().c_str(), *port);
+                    bool ret = Ssl::configureSSLUsingPkeyAndCertFromMemory(ssl, reply_message.getBody().c_str(), *port, signAlgorithm);
                     if (!ret)
                         debugs(33, 5, "Failed to set certificates to ssl object for PeekAndSplice mode");
                 } else {
@@ -4144,17 +4144,10 @@
     // Try to add generated ssl context to storage.
     if (port->generateHostCertificates && isNew) {
 
-        if (signAlgorithm == Ssl::algSignTrusted) {
-            // Add signing certificate to the certificates chain
-            X509 *cert = port->signingCert.get();
-            if (SSL_CTX_add_extra_chain_cert(sslContext, cert)) {
-                // increase the certificate lock
-                CRYPTO_add(&(cert->references),1,CRYPTO_LOCK_X509);
-            } else {
-                const int ssl_error = ERR_get_error();
-                debugs(33, DBG_IMPORTANT, "WARNING: can not add signing certificate to SSL context chain: " << ERR_error_string(ssl_error, NULL));
-            }
-            Ssl::addChainToSslContext(sslContext, port->certsToChain.get());
+        if (sslContext && (signAlgorithm == Ssl::algSignTrusted)) {
+            Ssl::chainCertificatesToSSLContext(sslContext, *port);
+        } else if (signAlgorithm == Ssl::algSignTrusted) {
+            debugs(33, DBG_IMPORTANT, "WARNING: can not add signing certificate to SSL context chain because SSL context chain is invalid!");
         }
         //else it is self-signed or untrusted do not attrach any certificate
 

=== modified file 'src/ssl/support.cc'
--- src/ssl/support.cc	2015-01-24 04:53:39 +0000
+++ src/ssl/support.cc	2015-04-30 04:00:12 +0000
@@ -1587,6 +1587,33 @@
     return createSSLContext(cert, pkey, port);
 }
 
+void
+Ssl::chainCertificatesToSSL(SSL *ssl, AnyP::PortCfg &port)
+{
+    SSL_CTX *sslContext = SSL_get_SSL_CTX(ssl);
+    if (sslContext) {
+        Ssl::chainCertificatesToSSLContext(sslContext, port);
+    } else {
+        debugs(33, DBG_IMPORTANT, "Could not retrieve SSL_CTX from SSL object");
+    }
+}
+
+void
+Ssl::chainCertificatesToSSLContext(SSL_CTX *sslContext, AnyP::PortCfg &port)
+{
+    assert(sslContext != NULL);
+    // Add signing certificate to the certificates chain
+    X509 *signingCert = port.signingCert.get();
+    if (SSL_CTX_add_extra_chain_cert(sslContext, signingCert)) {
+        // increase the certificate lock
+        CRYPTO_add(&(signingCert->references),1,CRYPTO_LOCK_X509);
+    } else {
+        const int ssl_error = ERR_get_error();
+        debugs(33, DBG_IMPORTANT, "WARNING: can not add signing certificate to SSL context chain: " << ERR_error_string(ssl_error, NULL));
+    }
+    Ssl::addChainToSslContext(sslContext, port.certsToChain.get());
+}
+
 bool
 Ssl::configureSSL(SSL *ssl, CertificateProperties const &properties, AnyP::PortCfg &port)
 {
@@ -1607,11 +1634,15 @@
     if (!SSL_use_PrivateKey(ssl, pkey.get()))
         return false;
 
+    if (properties.signAlgorithm == Ssl::algSignTrusted) {
+        Ssl::chainCertificatesToSSL(ssl, port);
+    }
+
     return true;
 }
 
 bool
-Ssl::configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::PortCfg &port)
+Ssl::configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::PortCfg &port, CertSignAlgorithm signAlgorithm)
 {
     Ssl::X509_Pointer cert;
     Ssl::EVP_PKEY_Pointer pkey;
@@ -1627,6 +1658,10 @@
     if (!SSL_use_PrivateKey(ssl, pkey.get()))
         return false;
 
+    if (signAlgorithm == Ssl::algSignTrusted) {
+        Ssl::chainCertificatesToSSL(ssl, port);
+    }
+
     return true;
 }
 

=== modified file 'src/ssl/support.h'
--- src/ssl/support.h	2015-01-13 09:13:49 +0000
+++ src/ssl/support.h	2015-04-30 04:00:12 +0000
@@ -222,6 +222,18 @@
 SSL_CTX * createSSLContext(Ssl::X509_Pointer & x509, Ssl::EVP_PKEY_Pointer & pkey, AnyP::PortCfg &port);
 
 /**
+ \ingroup ServerProtocolSSLAPI
+ * Chain signing certificate and chained certificates to an SSL Context
+ */
+void chainCertificatesToSSLContext(SSL_CTX *sslContext, AnyP::PortCfg &port);
+
+/**
+ \ingroup ServerProtocolSSLAPI
+ * Chain signing certificate and chained certificates to an SSL object
+ */
+void chainCertificatesToSSL(SSL *ssl, AnyP::PortCfg &port);
+
+/**
   \ingroup ServerProtocolSSLAPI
   * Generates a certificate and a private key using provided properies and set it
   * to SSL object.
@@ -233,7 +245,7 @@
   * Read private key and certificate from memory and set it to SSL object
   * using their.
  */
-bool configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::PortCfg &port);
+bool configureSSLUsingPkeyAndCertFromMemory(SSL *ssl, const char *data, AnyP::PortCfg &port, CertSignAlgorithm signAlgorithm);
 
 /**
   \ingroup ServerProtocolSSLAPI

