=== modified file 'src/http/Makefile.am'
--- src/http/Makefile.am	2015-09-05 18:52:17 +0000
+++ src/http/Makefile.am	2015-09-21 16:03:15 +0000
@@ -35,7 +35,4 @@
 	($(AWK) -f $(top_srcdir)/src/mk-string-arrays.awk sbuf=1 < $(srcdir)/MethodType.h | \
 		sed -e 's%METHOD_%%' -e 's%_C%-C%' >$@) || ($(RM) -f $@ && exit 1)
 
-RegisteredHeadersHash.cci: RegisteredHeadersHash.gperf
-	gperf --output-file=$@ -m 100000 $<
-
 CLEANFILES += MethodType.cc

=== modified file 'src/http/RegisteredHeadersHash.gperf'
--- src/http/RegisteredHeadersHash.gperf	2015-09-05 18:52:17 +0000
+++ src/http/RegisteredHeadersHash.gperf	2015-09-21 16:02:58 +0000
@@ -8,6 +8,11 @@
  * contributions from numerous individuals and organizations.
  * Please see the COPYING and CONTRIBUTORS files for details.
  */
+
+/* to regenerate:
+gperf --output-file=RegisteredHeadersHash.cci -m 100000 RegisteredHeadersHash.gperf
+remember to commit the generated file
+*/
 %}
 %language=C++
 %define hash-function-name HttpHeaderHash

