=== modified file 'src/cf.data.pre'
--- src/cf.data.pre	2016-11-19 07:24:46 +0000
+++ src/cf.data.pre	2016-11-20 20:02:19 +0000
@@ -5577,7 +5577,9 @@
 	will be considered fresh.
 
 	'Max' is an upper limit on how long objects without an explicit
-	expiry time will be considered fresh.
+	expiry time will be considered fresh. The value is also used
+	to form Cache-Control: max-age header for a request sent from
+	Squid to origin/parent.
 
 	options: override-expire
 		 override-lastmod

