=== modified file 'compat/os/linux.h'
--- compat/os/linux.h   2015-01-13 07:25:36 +0000
+++ compat/os/linux.h   2015-09-22 15:15:02 +0000
@@ -43,6 +43,13 @@
 typedef uint32_t __u32;
 #endif
 
+/* netinet/in.h must be included before any other sys header in order to properly
+ * activate include guards in <linux/libc-compat.h>
+ */
+#if HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
+
 #include <sys/capability.h>
 #endif /* HAVE_SYS_CAPABILITY_H */
 

