=== modified file 'src/HttpRequest.cc'
--- src/HttpRequest.cc	2016-11-04 16:47:34 +0000
+++ src/HttpRequest.cc	2016-12-07 19:40:41 +0000
@@ -545,6 +545,8 @@
 
         // XXX: this would seem the correct place to detect request cache-controls
         //      no-store, private and related which block cacheability
+        if (!flags.ignoreCc && cache_control && cache_control->noStore())
+            return false;
         break;
 
     case AnyP::PROTO_GOPHER:

