=== modified file 'scripts/source-maintenance.sh'
--- scripts/source-maintenance.sh	2015-08-04 05:32:35 +0000
+++ scripts/source-maintenance.sh	2015-09-22 01:58:16 +0000
@@ -277,6 +277,9 @@
 echo " "
 )| sed s%${ROOT}/src/%%g >${ROOT}/src/tests/Stub.list
 
+# Build the GPERF generated content
+make -C src/http gperf-files
+
 # Run formating
 echo "" >${ROOT}/doc/debug-sections.tmp
 srcformat || exit 1

=== modified file 'src/http/Makefile.am'
--- src/http/Makefile.am	2015-09-05 18:52:17 +0000
+++ src/http/Makefile.am	2015-09-22 02:12:05 +0000
@@ -21,7 +21,6 @@
 	RegisteredHeaders.h \
 	RegisteredHeaders.cc \
 	RegisteredHeadersHash.cci \
-	RegisteredHeadersHash.gperf \
 	RequestMethod.cc \
 	RequestMethod.h \
 	StatusCode.cc \
@@ -35,7 +34,11 @@
 	($(AWK) -f $(top_srcdir)/src/mk-string-arrays.awk sbuf=1 < $(srcdir)/MethodType.h | \
 		sed -e 's%METHOD_%%' -e 's%_C%-C%' >$@) || ($(RM) -f $@ && exit 1)
 
-RegisteredHeadersHash.cci: RegisteredHeadersHash.gperf
-	gperf --output-file=$@ -m 100000 $<
-
 CLEANFILES += MethodType.cc
+
+gperf-files: RegisteredHeadersHash.gperf
+	if test $(srcdir)/RegisteredHeadersHash.gperf -nt $(srcdir)/RegisteredHeadersHash.cci; then gperf --output-file=$(srcdir)/RegisteredHeadersHash.cci -m 100000 $(srcdir)/RegisteredHeadersHash.gperf; fi
+
+EXTRA_DIST = RegisteredHeadersHash.gperf
+
+.PHONY: gperf-files

