=== modified file 'src/mem/AllocatorProxy.h'
--- src/mem/AllocatorProxy.h	2015-08-24 17:53:25 +0000
+++ src/mem/AllocatorProxy.h	2015-09-30 11:30:22 +0000
@@ -35,7 +35,10 @@ class MemPoolMeter;
         assert(byteCount == sizeof(CLASS)); \
         return Pool().alloc(); \
     } \
-    void operator delete(void *address) {Pool().freeOne(address);} \
+    void operator delete(void *address) { \
+        if (address) \
+            Pool().freeOne(address); \
+    } \
     private:
 
 namespace Mem

