--- src/acl/Checklist.cc	2016-10-29 22:09:06.143778825 +0500
+++ src/acl/Checklist.cc	2016-10-29 22:09:49.380604810 +0500
@@ -397,7 +397,7 @@
 ACLChecklist::bannedAction(const allow_t &action) const
 {
     const bool found = std::find(bannedActions_.begin(), bannedActions_.end(), action) != bannedActions_.end();
-    debugs(28, 5, "Action '" << action << "/" << action.kind << (found ? " is " : "is not") << " banned");
+    debugs(28, 5, "Action '" << action << "/" << action.kind << (found ? "' is " : "' is not") << " banned");
     return found;
 }
 
