=== modified file 'src/http/Makefile.am'
--- src/http/Makefile.am	2015-09-22 12:00:37 +0000
+++ src/http/Makefile.am	2015-09-22 16:24:02 +0000
@@ -37,7 +37,7 @@
 CLEANFILES += MethodType.cc
 
 gperf-files: RegisteredHeadersHash.gperf
-	if test $(srcdir)/RegisteredHeadersHash.gperf -nt $(srcdir)/RegisteredHeadersHash.cci; then gperf --output-file=$(srcdir)/RegisteredHeadersHash.cci -m 100000 $(srcdir)/RegisteredHeadersHash.gperf; fi
+	if test $(srcdir)/RegisteredHeadersHash.gperf -nt $(srcdir)/RegisteredHeadersHash.cci; then gperf -m 100000 $(srcdir)/RegisteredHeadersHash.gperf | sed 's/register //g' > $(srcdir)/RegisteredHeadersHash.cci; fi
 
 EXTRA_DIST = RegisteredHeadersHash.gperf
 

=== modified file 'src/http/RegisteredHeadersHash.cci'
--- src/http/RegisteredHeadersHash.cci	2015-09-06 00:12:11 +0000
+++ src/http/RegisteredHeadersHash.cci	2015-09-22 16:25:19 +0000
@@ -81,7 +81,7 @@
 #ifndef GPERF_CASE_MEMCMP
 #define GPERF_CASE_MEMCMP 1
 static int
-gperf_case_memcmp (register const char *s1, register const char *s2, register unsigned int n)
+gperf_case_memcmp (const char *s1, const char *s2, unsigned int n)
 {
     for (; n > 0;)
     {
@@ -107,7 +107,7 @@
 };
 
 inline unsigned int
-HttpHeaderHashTable::HttpHeaderHash (register const char *str, register unsigned int len)
+HttpHeaderHashTable::HttpHeaderHash (const char *str, unsigned int len)
 {
     static const unsigned char asso_values[] =
     {
@@ -138,7 +138,7 @@
         114, 114, 114, 114, 114, 114, 114, 114, 114, 114,
         114, 114, 114, 114, 114, 114
     };
-    register int hval = len;
+    int hval = len;
 
     switch (hval)
     {
@@ -360,16 +360,16 @@
 };
 
 const struct HeaderTableRecord *
-HttpHeaderHashTable::lookup (register const char *str, register unsigned int len)
+HttpHeaderHashTable::lookup (const char *str, unsigned int len)
 {
     if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
     {
-        register int key = HttpHeaderHash (str, len);
+        int key = HttpHeaderHash (str, len);
 
         if (key <= MAX_HASH_VALUE && key >= 0)
             if (len == lengthtable[key])
             {
-                register const char *s = HttpHeaderDefinitionsTable[key].name;
+                const char *s = HttpHeaderDefinitionsTable[key].name;
 
                 if ((((unsigned char)*str ^ (unsigned char)*s) & ~32) == 0 && !gperf_case_memcmp (str, s, len))
                     return &HttpHeaderDefinitionsTable[key];

