=== modified file 'configure.ac'
--- configure.ac	2017-01-01 00:12:22 +0000
+++ configure.ac	2017-01-07 08:19:06 +0000
@@ -381,37 +381,14 @@
                   Optimization is good for production builds, but not
                   good for debugging. During development, use
                   --disable-optimizations to reduce compilation times
-                  and allow easier debugging. This option implies
-                  --disable-inline]), [
+                  and allow easier debugging.]), [
 if test "x$enableval" = "xno" ; then
   AC_MSG_NOTICE([Disabling compiler optimizations (-O flag)])
   CFLAGS="`echo $CFLAGS | sed -e 's/-O[[0-9g]]*//'`"
   CXXFLAGS="`echo $CXXFLAGS | sed -e 's/-O[[0-9g]]*//'`"
-  enable_inline="no"
 fi
 ])
 
-AH_TEMPLATE(_USE_INLINE_,[Include inline methods into header file])
-AC_ARG_ENABLE(inline,
-  AS_HELP_STRING([--disable-inline],
-                 [Do not compile trivial methods as inline. Squid
-                  is coded with much of the code able to be inlined.
-                  Inlining is good for production builds, but not
-                  good for development. During development, use
-                  --disable-inline to reduce compilation times and
-                  allow incremental builds to be quick. For
-                  production builds, or load tests, use
-                  --enable-inline to have squid make all trivial
-                  methods inlinable by the compiler.]), [
-SQUID_YESNO([$enableval],
-  [Unrecognized argument to --disable-inline: $enableval])
-])
-AC_MSG_NOTICE([inlining optimizations enabled: ${enable_inline:=yes}])
-SQUID_DEFINE_BOOL(_USE_INLINE_,$enable_inline,
-  [Include inline methods into header file])
-# to be used by sub-commands
-export enable_inline
-
 AC_ARG_ENABLE(debug-cbdata,
   AS_HELP_STRING([--enable-debug-cbdata],
       [Provide some debug information in cbdata]), [ 

=== modified file 'doc/release-notes/release-5.sgml'
--- doc/release-notes/release-5.sgml	2017-01-06 15:09:29 +0000
+++ doc/release-notes/release-5.sgml	2017-01-07 08:21:36 +0000
@@ -121,7 +121,8 @@
 <sect1>Changes to existing options<label id="modifiedoptions">
 <p>
 <descrip>
-	<p>No changes to existing options in this version.
+	<tag>--disable-optimizations</tag>
+	<p>No longer implies <em>--disable-inline</em> option (which is removed).
 
 </descrip>
 </p>
@@ -129,7 +130,8 @@
 <sect1>Removed options<label id="removedoptions">
 <p>
 <descrip>
-	<p>No removed options in this version.
+	<tag>--disable-inline</tag>
+	<p>Removed.
 
 </descrip>
 

=== modified file 'include/squid.h'
--- include/squid.h	2017-01-01 00:12:22 +0000
+++ include/squid.h	2017-01-07 08:18:04 +0000
@@ -28,12 +28,6 @@
 #define SQUIDCEXTERN extern
 #endif
 
-#if _USE_INLINE_
-#define _SQUID_INLINE_ inline
-#else
-#define _SQUID_INLINE_
-#endif
-
 /****************************************************************************
  *--------------------------------------------------------------------------*
  * DO *NOT* MAKE ANY CHANGES below here unless you know what you're doing...*

=== modified file 'test-suite/buildtests/layer-01-minimal.opts'
--- test-suite/buildtests/layer-01-minimal.opts	2017-01-01 00:12:22 +0000
+++ test-suite/buildtests/layer-01-minimal.opts	2017-01-07 08:22:41 +0000
@@ -39,7 +39,6 @@
 	--disable-loadable-modules \
 	--disable-gnuregex \
 	--disable-optimizations \
-	--disable-inline \
 	--disable-debug-cbdata \
 	--disable-xmalloc-statistics \
 	--disable-async-io \

=== modified file 'test-suite/buildtests/layer-02-maximus.opts'
--- test-suite/buildtests/layer-02-maximus.opts	2017-01-01 00:12:22 +0000
+++ test-suite/buildtests/layer-02-maximus.opts	2017-01-07 08:22:50 +0000
@@ -54,7 +54,6 @@
 	--enable-loadable-modules \
 	--enable-gnuregex \
 	--enable-optimizations \
-	--enable-inline \
 	--enable-debug-cbdata \
 	--enable-xmalloc-statistics \
 	--enable-async-io \

